Issue 857: Fix bug that can prevent linting ephemeral models

barrywhart opened the pull request:
### Brief summary of the change made

Fixes #857

Adds a new public method to the templater interface to let the templater determine the order for linting files. With dbt, this is used to process dependent models (ephemeral models in particular!) before the models that use them.

### Are there any other side effects of this change that we should be aware of?

* Extends the public interface for templaters (`RawTemplater` and its descendants)
* Changes two more properties of `DbtTemplater` to use `@cached_property` (`profiles_dir` and `project_dir`)
### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/parser` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

deanja commented:
Hi Barry @barrywhart 

The fix worked in my `b.sql` `c.sql` test fixture, thank you.   But I am seeing some odd behaviour in my real-world dbt project.

Note that I'm new to python packages and testing PRs so I might have done something odd in the setup.  I accessed your PR code by doing this in Poetry:
`$ poetry remove sqlfluff`

Add to pyproject.toml:
`sqlfluff = { git = "https://github.com/barrywhart/sqlfluff.git", branch = "bhart-issue_857_unable_to_lint_ephemeral_model" }`

`$ poetry update`

The odd behaviour in my real-world dbt project is:

`sqlfluff lint` does not FAIL files that break the rules.

`sqlfluff lint` not longer outputs these lines at the beginning of the linting run:
```
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
```

Investigation by linting specific models shows shows some new behaviour.

- `sqlfluff lint path/to/model/my_model.sql` throws an error: `KeyError: 'source.metering.dw1_dimensions.dim_entity'`. Where that source is used in the linted model or an upstream model.
- `sqlfluff lint path/to/model/my_other_model` returns sqlfluff FAIL of an upstream model, returning message `Found unparseable section`.  It is trying to parse code from way back in my external_stage dbt macro.

The project is on dbt 0.20.2 .  dbt compile, run, test are all working as expected.
It's bit of work to factor those models and macros out of my real-world project to share. I can dig deeper if you have any suggestions.

barrywhart commented:
@deanja: The `KeyError` sounds like it's related to this PR. It probably happens because a model in the project references it, and now SQLFluff wants to process that model first. Do you have any ideas why that model is not part of the project? For now, I will change the code to skip cases like this rather than failing, but I'm a bit concerned that is masking a problem rather than fixing it.

IIUC, the "unparseable" error sounds like correct behavior with respect to **this** PR: your project has some code that SQLFluff can't parse. Can you try and see what is causing the parse error? You may want to create a separate issue for that.

barrywhart commented:
Ok, I think the `KeyError` should be addressed now. In order to help understand how common this is, I added a log warning message to the dbt templater, e.g.
```
Dependent model source.metering.dw1_dimensions.dim_entity not found in project. Skipping.
```

barrywhart commented:
I'll look into the first two things also.

barrywhart commented:
Ok, I fixed the issue that was causing some output not to be seen, e.g. the "dbt templater" messages you mentioned:
```
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
```

About this:
>`sqlfluff lint` does not FAIL files that break the rules.

This is the behavior I'm seeing after the fix above. When you say "it does not FAIL fails", is this what you meant? Does it work better now? If not, what output are you seeing?
```
(sqlfluff-3.7.9) ➜  my-dbt sqlfluff lint models/error
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
== [models/error/c.sql] FAIL
L:   3 | P:   1 | L010 | Keywords must be upper case.
L:   3 | P:  10 | L010 | Keywords must be upper case.
== [models/error/b.sql] FAIL
L:   1 | P:   1 | L010 | Keywords must be upper case.
L:   1 | P:   9 | L001 | Unnecessary trailing whitespace.
L:   2 | P:   1 | L010 | Keywords must be upper case.
L:   3 | P:   1 | L010 | Keywords must be upper case.
L:   3 | P:   7 | L014 | Unquoted identifiers must be consistently capitalised.
All Finished 📜 🎉!
(sqlfluff-3.7.9) ➜  my-dbt echo $?
65
```

deanja commented:
Bedtime, here in Australia, but here's some of the info you asked for.  I
can take a deeper look in a day or two.

"sqlfluff lint does not FAIL files that break the rules."
means the command's only output was "All finished" or similar, even though
the sql files have sql that breaks the linting rules.

`source.metering.dw1_dimensions.dim_entity`   is a dbt Source, defined in
my staging/sources.yml.  `dbt compile` and `dbt run` both work with that
Source, without error.  I don't know why it would throw a KeyError.

The `unparseable` section error was for an upstream model of the one I
specifically linted.   That upstream model calls a macro that may have some
funkiness in it - a known bit of tech debt in that dbt project. It may be
better for me to update the macro than raise a PR for you ::).

What surprised me about the `unparseable` was that  sqlfluff went there
even though that upstream model's folder is excluded in
.sqlfluffignore:with:
```
# Avoid problem where sqlfluff can't parse the external_stage macro.
sqlfluff 0.4.1
# Was unable to stub the macro using sqlfluff:templater config in a
.sqlfluff file
# in  the models/raw subdirectory.
models/raw/
```

I'm wondering if this PR enables sqlfluff to traverse models that it
previously would not have?  Perhaps it's become more 'curious' in its
effort to resolve `ephemeral` models.   Sorry if I opened a can of worms
for you.




On Mon, 4 Oct 2021 at 22:58, Barry Hart ***@***.***> wrote:

> Ok, I fixed the issue that was causing some output not to be seen, e.g.
> the "dbt templater" messages you mentioned:
>
> === [dbt templater] Compiling dbt project...
>
> === [dbt templater] Project Compiled.
>
>
> About this:
>
> sqlfluff lint does not FAIL files that break the rules.
>
> This is the behavior I'm seeing after the fix above. When you say "it does
> not FAIL fails", is this what you meant? Does it work better now? If not,
> what output are you seeing?
>
> (sqlfluff-3.7.9) ➜  my-dbt sqlfluff lint models/error
>
> === [dbt templater] Compiling dbt project...
>
> === [dbt templater] Project Compiled.
>
> == [models/error/c.sql] FAIL
>
> L:   3 | P:   1 | L010 | Keywords must be upper case.
>
> L:   3 | P:  10 | L010 | Keywords must be upper case.
>
> == [models/error/b.sql] FAIL
>
> L:   1 | P:   1 | L010 | Keywords must be upper case.
>
> L:   1 | P:   9 | L001 | Unnecessary trailing whitespace.
>
> L:   2 | P:   1 | L010 | Keywords must be upper case.
>
> L:   3 | P:   1 | L010 | Keywords must be upper case.
>
> L:   3 | P:   7 | L014 | Unquoted identifiers must be consistently capitalised.
>
> All Finished 📜 🎉!
>
> (sqlfluff-3.7.9) ➜  my-dbt echo $?
>
> 65
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sqlfluff/sqlfluff/pull/1496#issuecomment-933411774>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANX27CRCX3DW4E4RVHXPYU3UFGJGLANCNFSM5FHZKERA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

barrywhart commented:
>"sqlfluff lint does not FAIL files that break the rules."
>means the command's only output was "All finished" or similar, even though
>the sql files have sql that breaks the linting rules.

Please try this again after my fix. Or if you've done that already, can you test it on the "error" test you provided in the issue? I provided my output from that one above, and the lint issues are seen in the output.

About the `.sqlfluffignore`, yes, this new code is not respecting that. I'll look into that.

>Sorry if I opened a can of worms for you.

Ha, I think we are both in a can of worms. But not surprising. The more features a tool has, the more interactions there are. That's why it's really helpful that you are testing this on a real project. I am new to this area of the code, so it's easy to be unaware of something important. A working algorithm is only a start...

deanja commented:
I tried out sqlfluff (0.6.6 72e7a79)

The nice messages are back:

=== [dbt templater] Compiling dbt project...

=== [dbt templater] Project Compiled.

Successful linting of a trivial model in my real-world project:

dw-wip/metering/dbt/models/example/my_first_dbt_model.sql] FAIL
L:   9 | P:   1 | L010 | Keywords must be upper case.

But it is still failing to enforce similar rules in other models.  Eg the
one with the
select    * FROM {{ source('dw1_dimensions', 'dim_entity') }}
That's the one that was causing the KeyError.

On that aspect, sorry for newb question, but how do I run sqlfluff so as to
see the warning message, like this one you described?:

Dependent model source.metering.dw1_dimensions.dim_entity not found in
project. Skipping.

I see that warning when linting the model specifically, but not with the
general `sqlfluff lint` of all models in the project.

Have to stop for the night now.   My thoroughness for the day has been used
up :)

On Mon, 4 Oct 2021 at 23:43, Barry Hart ***@***.***> wrote:

> "sqlfluff lint does not FAIL files that break the rules."
> means the command's only output was "All finished" or similar, even though
> the sql files have sql that breaks the linting rules.
>
> Please try this again after my fix. Or if you've done that already, can
> you test it on the "error" test you provided in the issue? I provided my
> output from that one above, and the lint issues are seen in the output.
>
> About the .sqlfluffignore, yes, this new code is not respecting that.
> I'll look into that.
>
> Sorry if I opened a can of worms for you.
>
> Ha, I think we are both in a can of worms. But not surprising. The more
> features a tool has, the more interactions there are. That's why it's
> really helpful that you are testing this on a real project. I am new to
> this area of the code, so it's easy to be unaware of something important. A
> working algorithm is only a start...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sqlfluff/sqlfluff/pull/1496#issuecomment-933445936>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANX27CTZTSETPDBDCME2HTLUFGON5ANCNFSM5FHZKERA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

barrywhart commented:
Ok, I updated the code so it should now respect ".sqlfluffignore".

You shouldn't have to do anything special to see the new warning. If you're not seeing it, you could try adding `-vvvv` to the command. I don't think that should be needed.

About this:
>But it is still failing to enforce similar rules in other models.  Eg the
>one with the
>select    * FROM {{ source('dw1_dimensions', 'dim_entity') }}
>That's the one that was causing the KeyError.

Would it make sense to create a new issue for that? In other words, does this PR address the original issue #857, "Unable to lint Ephemeral model", well enough that it's worth including in a release? IIUC:
* The PR is now working "reasonably" well.
* Your testing turned up several issues.
  * The most severe (& closely related) issues are now fixed.
  * There's apparently a remaining issue where some files are not being linted.

Thanks for your work in testing this! I'll see if I can get some other dbt users to test this today, from North America or Europe. The sun never sets on SQLFluff. 😄

pwildenhain commented:
Ran this locally and all tests passed 🎉 👯‍♂️ 

Really nice work here 👌🏻
